### PR TITLE
ExternalAHRS: Add the ability to send commands to the EAHRS

### PR DESCRIPTION
**Pull request by valbv**
It's for the ability to send commands to EAHRS like "Enable/Disable GNSS", "Start/Stop magnitometer calibration in flight" and everything that device can receive

### CODE REVIEW

**Comment by Ryanf55** [comment]
Nice. If you can, feel free to add one such packet so that this code is used by a driver.

**Comment by valbv** [comment]
Linked PR in MAVLink repo:
https://github.com/ArduPilot/mavlink/pull/376

**Review by peterbarker — COMMENTED** [review]
I'd prefer to have something at the AHRS-backend level rather than at the EAHRS level for this.  Turning GNSS off is a thing for EKF3, for example, just not in master right now.

**Comment by valbv** [comment]
@peterbarker , where can I find the code you were talking about (Turning GNSS off for EKF3)?
I want to avoid conflicts in the names of methods.

I looked through the branches in the repository and requests, but I couldn't find it

**Comment by Ryanf55** [comment]
Check the code path of aux function 65 "GPS Disable": https://ardupilot.org/copter/docs/common-auxiliary-functions.html

**Comment by valbv** [comment]
Did I get the idea right, you want to see something like this?

File: libraries/AP_AHRS/AP_AHRS.h [hunk 1]
```diff
@@ -683,6 +683,9 @@ class AP_AHRS {
     // get access to an EKFGSF_yaw estimator
     const EKFGSF_yaw *get_yaw_estimator(void) const;
 
+    void set_data_send_disable(bool disable);
```
Comment (Ryanf55) [inline libraries/AP_AHRS/AP_AHRS.h:686]: The name isn't apparent what this does. Can you add a comment? 

Does this disable the EARHS from sending data back to the autopilot?

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS.h [hunk 2]
```diff
@@ -121,6 +121,9 @@ class AP_ExternalAHRS {
     bool get_accel(Vector3f &accel);
     void send_status_report(class GCS_MAVLINK &link) const;
     bool get_variances(float &velVar, float &posVar, float &hgtVar, Vector3f &magVar, float &tasVar) const;
+    void set_data_send_disable(bool disable);
+    void set_gnss_disable(bool disable);
```
Comment (Ryanf55) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS.h:125]:
```suggestion
    void set_gnss_disable(bool disable);
```

Does setting this to false re-enable? If so, thoughts on calling it `set_gnss_state(bool enabled)`?

File: libraries/AP_AHRS/AP_AHRS.h (hunk 1 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.h]: I added the comment

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS.h (hunk 2 above)
Comment (valbv) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS.h]: Yes, I renamed them to

```cpp
void set_data_sending_state(bool enabled);
void set_gnss_state(bool enabled)
```

**Comment by valbv** [comment]
Yes, we used `set_gnss_state` method-logic for the dead reckoning tests with EAHRS

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 3]
```diff
@@ -3629,6 +3629,30 @@ float AP_AHRS::get_air_density_ratio(void) const
...
+    }
+}
+
+void AP_AHRS::set_gnss_state(bool enabled)
+{
+    switch (active_EKF_type()) {
+#if AP_AHRS_EXTERNAL_ENABLED
+    case EKFType::EXTERNAL:
+        return external.set_gnss_state(enabled);
+#endif
```
Comment (Ryanf55) [inline libraries/AP_AHRS/AP_AHRS.cpp:3650]: The code paths that aux function 105 takes could be used to disable GPS in EKF2/3. 
[105](https://ardupilot.org/copter/docs/common-auxiliary-functions.html)

Not required for the PR.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 4]
```diff
@@ -4878,6 +4878,39 @@ MAV_RESULT GCS_MAVLINK::handle_command_request_autopilot_capabilities(const mavl
 }
 #endif
 
+#if AP_AHRS_ENABLED
+MAV_RESULT GCS_MAVLINK::handle_AHRS_message(const mavlink_command_int_t &packet)
+{
+    switch (packet.command) {
+        case MAV_CMD_AHRS_START:
+        {
+            AP::ahrs().set_data_sending_state(/*enabled*/ true);
```
Comment (Ryanf55) [inline libraries/GCS_MAVLink/GCS_Common.cpp:4887]: I know it's ON/OFF, but the fact you need a commen there may be a good reason we can do 

```c++
enum class GNSS_SENDING_STATE {
  ENABLED,
  DISABLED
};
```

**Review by Ryanf55 — APPROVED** [review]
Nice, yea I'm overall on board with the approach here. I haven't tested it yet.

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 3 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.cpp:3650]: Ok, I'll create a next request with these changes

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 4 above)
Comment (valbv) [reply libraries/GCS_MAVLink/GCS_Common.cpp:4887]: Yeah, I was thinking about that too.
If you don't mind, I'll start a separate next request with the "enum class" using

**Comment by peterbarker** [comment]
This is looking nicer than the write-bytes approach.

I'm going to mark this for DevCall for discussion.

**Comment by Ryanf55** [comment]
Nothing to add from me, I'm in favor of this change and believe it sets up a nice pattern for adding future generic commands to the EKF/EAHRS.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_InertialLabs.cpp [hunk 5]
```diff
@@ -1120,5 +1128,31 @@ bool AP_ExternalAHRS_InertialLabs::get_variances(float &velVar, float &posVar, f
...
 }
 
+bool AP_ExternalAHRS_InertialLabs::write_bytes(const char *bytes, uint8_t len)
+{
+    return uart->write(reinterpret_cast<const uint8_t *>(bytes), len);
+}
+
+void AP_ExternalAHRS_InertialLabs::set_data_sending_state(bool enabled)
+{
+    if (enabled)
```
Comment (rmackay9) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_InertialLabs.cpp:1138]: nitpick: bracket should be on the same line as the if

**Comment by rmackay9** [comment]
Hi @valbv,

Thanks for this.  What's the purpose of disabling an EAHRS?  I guess it's to test a complete failure of the external AHRS system?  I wonder what testing has been done for this?

File: libraries/AP_AHRS/AP_AHRS_External.cpp [hunk 6]
```diff
@@ -137,4 +137,14 @@ void AP_AHRS_External::get_control_limits(float &ekfGndSpdLimit, float &ekfNavVe
     ekfNavVelGainScaler = 0.5;
 }
 
+void AP_AHRS_External::set_data_sending_state(bool enabled)
```
Comment (rmackay9) [inline libraries/AP_AHRS/AP_AHRS_External.cpp:140]: let's put some comments above the methods..

File: libraries/AP_AHRS/AP_AHRS_External.h [hunk 7]
```diff
@@ -87,6 +87,9 @@ class AP_AHRS_External : public AP_AHRS_Backend {
     void send_ekf_status_report(class GCS_MAVLINK &link) const override;
 
     void get_control_limits(float &ekfGndSpdLimit, float &controlScaleXY) const override;
+
+    void set_data_sending_state(bool enabled) override;
```
Comment (rmackay9) [inline libraries/AP_AHRS/AP_AHRS_External.h:91]: comments here too would be nice

**Comment by rmackay9** [comment]
More comments would be nice, there are very few which makes it difficult to review but also it will be even more difficult for future developers trying to figure out how to use the various methods

File: libraries/AP_AHRS/AP_AHRS_External.cpp (hunk 6 above)
Comment (Ryanf55) [reply libraries/AP_AHRS/AP_AHRS_External.cpp:140]: To elaborate, at minimum, copy the comments across all the headers. Although the clasess inherit/override methods, not everyone's editor propogates the description, so AP generally copies comments.

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 3 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.cpp:3670]: I'm not sure that I understood what I need right, but I added the GPS disabling for EKF2/3

File: libraries/AP_AHRS/AP_AHRS.h (hunk 1 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.h]: Done

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 4 above)
Comment (valbv) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Done

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS.cpp [hunk 8]
```diff
@@ -375,6 +375,35 @@ void AP_ExternalAHRS::send_status_report(GCS_MAVLINK &link) const
...
+}
+
+bool AP_ExternalAHRS::write_bytes(const char *bytes, uint8_t len)
+{
+    if (!backend) {
+        return false;
+    }
+    return backend->write_bytes(bytes, len);
+}
+
```
[the remark this answers is not in the document]
Comment (valbv) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS.cpp]: Done

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS.h [hunk 9]
```diff
@@ -121,6 +121,9 @@ class AP_ExternalAHRS {
     bool get_accel(Vector3f &accel);
     void send_status_report(class GCS_MAVLINK &link) const;
     bool get_variances(float &velVar, float &posVar, float &hgtVar, Vector3f &magVar, float &tasVar) const;
+    void set_data_sending_state(bool enabled);
+    void set_gnss_state(bool enabled);
+    bool write_bytes(const char *bytes, uint8_t len);
```
[the remark this answers is not in the document]
Comment (valbv) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS.h]: Done

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_InertialLabs.h [hunk 10]
```diff
@@ -220,6 +220,9 @@ class AP_ExternalAHRS_InertialLabs : public AP_ExternalAHRS_backend {
     void update_thread();
     bool check_uart();
     bool check_header(const ILabsHeader *h) const;
+    bool write_bytes(const char *bytes, uint8_t len) override;
```
[the remark this answers is not in the document]
Comment (valbv) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS_InertialLabs.h]: Done

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_backend.h [hunk 11]
```diff
@@ -42,6 +42,9 @@ class AP_ExternalAHRS_backend {
     virtual bool pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const = 0;
     virtual void get_filter_status(nav_filter_status &status) const {}
     virtual bool get_variances(float &velVar, float &posVar, float &hgtVar, Vector3f &magVar, float &tasVar) const { return false; }
+    virtual void set_data_sending_state(bool enabled) {}
+    virtual void set_gnss_state(bool enabled) {}
+    virtual bool write_bytes(const char *bytes, uint8_t len) { return false; }
```
[the remark this answers is not in the document]
Comment (valbv) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS_backend.h]: Done

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 12]
```diff
@@ -4878,6 +4878,39 @@ MAV_RESULT GCS_MAVLINK::handle_command_request_autopilot_capabilities(const mavl
 }
 #endif
 
+#if AP_AHRS_ENABLED
+MAV_RESULT GCS_MAVLINK::handle_AHRS_message(const mavlink_command_int_t &packet)
+{
+    switch (packet.command) {
+        case MAV_CMD_AHRS_START:
+        {
```
[the remark this answers is not in the document]
Comment (valbv) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Done

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_InertialLabs.cpp (hunk 5 above)
Comment (valbv) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS_InertialLabs.cpp]: Done

File: libraries/AP_AHRS/AP_AHRS_External.cpp (hunk 6 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS_External.cpp]: Done

File: libraries/AP_AHRS/AP_AHRS_External.h (hunk 7 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS_External.h]: Done

**Comment by rmackay9** [comment]
Are we sure that we want the new methods to be in AP_AHRS?  We could reduce the scope by only making them accessible via AP_ExternalAHRS

**Comment by Ryanf55** [comment]
> Are we sure that we want the new methods to be in AP_AHRS? We could reduce the scope by only making them accessible via AP_ExternalAHRS

Reasoning: https://github.com/ArduPilot/ardupilot/pull/28485#pullrequestreview-2486838660

**Comment by rmackay9** [comment]
I wonder if we even really need the mavlink command?  Some of our other subsystems can be enabled/disabled using a _TYPE parameter.  If it's not something that users will regularly do a parameter can be more convenient and accessible (most GCSs won't support this mavlink command I suspect)

**Comment by valbv** [comment]
If this PR will be Ok, I plan to add common commands for AHRS In the next PRs:
- Start/Stop magnetometer calibration
- Send aiding data

That's a reason why I chose the MAVLink protocol extension.
And then add the "AHRS" tab to the Mission Planner

**Comment by tridge** [comment]
@valbv for this type of runtime configuration we have in the past used aux functions, which are available over mavlink as well as R/C switches and lua scripts. Is there some reason for not continuing with that approach? 
I also wouldn't mind having a COMMAND_INT that is reserved for external AHRS, where the interpretation of the params is up to each backend. That would give a lot of flexibility if aux functions aren't sufficient

**Comment by valbv** [comment]
@tridge, sorry for the long absence.
Aux functions seem to be quite sufficient for this functionality.
I removed changes in MAVLink and added RC-Channel processing for it.

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 13]
```diff
@@ -3652,6 +3652,50 @@ float AP_AHRS::get_air_density_ratio(void) const
     return 1.0 / sq(eas2tas);
 }
 
+void AP_AHRS::set_data_sending_state(DATA_SENDING_STATE data_sending_state)
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS.cpp:3655]:
```suggestion
void AP_AHRS::set_data_sending_state(DataSendingState new_state)
```

similarly for other enum names

File: libraries/AP_AHRS/AP_AHRS_Backend.h [hunk 14]
```diff
@@ -292,4 +292,17 @@ class AP_AHRS_Backend
     virtual void set_terrain_hgt_stable(bool stable) {}
 
     virtual void get_control_limits(float &ekfGndSpdLimit, float &controlScaleXY) const = 0;
+
+    enum class DATA_SENDING_STATE {
+        ENABLED,
+        DISABLED
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_Backend.h:298]:
```suggestion
        DISABLED,
```

similarly elsewhere

File: libraries/AP_AHRS/AP_AHRS_Backend.h [hunk 15]
```diff
@@ -292,4 +292,17 @@ class AP_AHRS_Backend
...
+    };
+    // enable/disable sending data to autopilot. Can be used to simulate AHRS failure.
+    virtual void set_data_sending_state(DATA_SENDING_STATE state) {};
+
+    enum class GPS_STATE {
+        ENABLED,
+        DISABLED
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_Backend.h:305]:
```suggestion
        DISABLED,
```

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 16]
```diff
@@ -3652,6 +3652,50 @@ float AP_AHRS::get_air_density_ratio(void) const
...
+                                               AP_AHRS_Backend::DATA_SENDING_STATE::ENABLED);
+#endif
+    default:
+        break;
+    }
+}
+
+void AP_AHRS::set_gps_state(GPS_STATE gps_state)
+{
+    switch (active_EKF_type()) {
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS.cpp:3671]: This is somewhat odd.  You're turning things off the for active EKLF type.  Presumably it's not going to stay the active EKF type for long, so turning the functionality back on again could be rather problematic?

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 17]
```diff
@@ -3652,6 +3652,50 @@ float AP_AHRS::get_air_density_ratio(void) const
...
+#endif //HAL_NAVEKF3_AVAILABLE
+
+#if AP_AHRS_EXTERNAL_ENABLED
+    case EKFType::EXTERNAL:
+        return external.set_gps_state(gps_state == GPS_STATE::DISABLED ?
+                                      AP_AHRS_Backend::GPS_STATE::DISABLED :
+                                      AP_AHRS_Backend::GPS_STATE::ENABLED);
+#endif
+    default:
+        break;
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS.cpp:3695]: Avoid the use of default: clauses, spell out each backend type

File: libraries/AP_AHRS/AP_AHRS_External.cpp [hunk 18]
```diff
@@ -137,4 +137,18 @@ void AP_AHRS_External::get_control_limits(float &ekfGndSpdLimit, float &ekfNavVe
     ekfNavVelGainScaler = 0.5;
 }
 
+void AP_AHRS_External::set_data_sending_state(DATA_SENDING_STATE state)
+{
+    AP::externalAHRS().set_data_sending_state(state == DATA_SENDING_STATE::ENABLED ?
+                                              AP_ExternalAHRS::DATA_SENDING_STATE::ENABLED :
+                                              AP_ExternalAHRS::DATA_SENDING_STATE::DISABLED);
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_External.cpp:144]:
```suggestion
    const bool ext_state = (state == DataSendingState::ENABLED) ?
                                              AP_ExternalAHRS::DataSendingState::ENABLED :
                                              AP_ExternalAHRS::DataSendingState::DISABLED
    AP::externalAHRS().set_data_sending_state(ext_state);
```

similarly elsewhere

**Review by peterbarker — COMMENTED** [review]
I'm marking this for DevCall.

Looking at this, I think the AHRS interface is problematic and will remain so until we clean up the AHRS backend stuff properly.

I'm thinking one of two things:
 - explicit aux functions for disabling stuff in ExternalAHRS (ie. calls directly on the EAHRS library
 - passing in a "dummy" instance number for the AHRS enable/disable calls, 0 for primary, 1 for secondary estimators.  You'd need separate functions for each.
 -

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 13 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.cpp]: Done

File: libraries/AP_AHRS/AP_AHRS_Backend.h (hunk 14 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS_Backend.h]: Done

File: libraries/AP_AHRS/AP_AHRS_Backend.h (hunk 15 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS_Backend.h]: Done

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 16 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.cpp]: Fixed

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 17 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.cpp]: Done

File: libraries/AP_AHRS/AP_AHRS_External.cpp (hunk 18 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS_External.cpp]: Done

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 19]
```diff
@@ -3652,6 +3652,34 @@ float AP_AHRS::get_air_density_ratio(void) const
...
+    external.set_gps_state(ext_state);
+#endif
+
+#if HAL_NAVEKF2_AVAILABLE && AP_GPS_ENABLED
+    AP::gps().force_disable(new_state == GpsState::DISABLED);
+#endif
+
+#if HAL_NAVEKF3_AVAILABLE && AP_GPS_ENABLED
+    AP::gps().force_disable(new_state == GpsState::DISABLED);
+#endif
```
Comment (IamPete1) [inline libraries/AP_AHRS/AP_AHRS.cpp:3680]: Don't need this twice.

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 20]
```diff
@@ -3652,6 +3652,34 @@ float AP_AHRS::get_air_density_ratio(void) const
...
+{
+#if AP_AHRS_EXTERNAL_ENABLED
+    const AP_AHRS_Backend::DataSendingState ext_state = (new_state == DataSendingState::DISABLED) ?
+                                                            AP_AHRS_Backend::DataSendingState::DISABLED :
+                                                            AP_AHRS_Backend::DataSendingState::ENABLED;
+    external.set_data_sending_state(ext_state);
+#endif
+}
+
+void AP_AHRS::set_gps_state(GpsState new_state)
```
Comment (IamPete1) [inline libraries/AP_AHRS/AP_AHRS.cpp:3665]: Is this method used?

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 19 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.cpp:3680]: Sorry, I don't understand what you mean. Can you rephrase it, please?

Here are cases for EKF2 and EKF3

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 20 above)
Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.cpp:3665]: Yes, in the "RC_Channel.cpp"

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 19 above)
Comment (IamPete1) [reply libraries/AP_AHRS/AP_AHRS.cpp:3680]: Right, but you can combine the defines. 
```
#if (HAL_NAVEKF2_AVAILABLE || HAL_NAVEKF3_AVAILABLE) && AP_GPS_ENABLED
```

No need to duplicate the function call.

Comment (valbv) [reply libraries/AP_AHRS/AP_AHRS.cpp]: Done
